### ISSUE TITLE
Be able to re-use windows/splits by plugins

### ISSUE THREAD

**Issue by jgclark** [NotePlan/plugins#370]
We can ask for new windows or split-view panels to be shown from API calls.
But as we use these more, we need some other controls as well. Particularly, to reuse existing splits/windows if open when opening a new split or window. This will stop many windows piling up when calling the same commands in quick succession.

I can think of various potential solutions, but as I don't know the relative complexity of these, I'll restrict myself to giving some more detailed use cases.

### SearchExtensions' /quickSearch and /searchOver* commands
These commands open a new split window as a convenience when they run. When you run the command again (or via the psuedo-button at the top of the note), another split window is created, even when there's an existing split showing the same note.
I want a way to say "show note in new split, unless that note is already open in a window, in which case just give it focus".

### Reviews' /project lists command
This opens an HTML window. If it's run again, it re-uses the same window -- this is what I want in the first use case above.
But it's easier here, as there's a limit of just 1 HTML window at a time.
When that limit is removed, this also needs a way to say 'reuse'.

### NoteHelpers' /open current note new split
To make things slightly harder, this command shows the need to be able to deliberately open a second split view of the same note as is already open in the Editor.

**Comment by dwertheimer** [comment]
Yes, this is much-needed. Here's an idea. Assuming there can only be one Editor at at time:

```
Editor.windows() => 
[ 0: { id:XXX type: "main", note: {TNote}
  1: { id: YYY type: "split", note: {TNote}
  2: { id: ZZZ type: "floating", note: {TNote} ]
 
```
`const note = Editor.focusWindow(id) `// brings it to front, makes it Editor, as if someone clicked on it, returns the TNote of the window focused so you can double-check that it's the one you want (or null/false? if it was not able to find the window ID you asked for)

 In this way, we could write a helper like this:
 ```js
 const getWindowWithFilename = (filename:string) => Editor.windows.find(w=>w.note.filename === filename)
 const focusWindowWithFilename = (filename:string) => Editor.focusWindow(getWindowWithFilename(filename))
 ```
 
 I'd also like to pose a question...should you be able to *close* a split or floating window via plugin? I think you should. In that case, the companion to ` Editor.focusWindow(id) ` would be `  Editor.closeWindow(id) `

**Comment by jgclark** [comment]
Thanks, @dwertheimer, for reviewing this quickly. As you see I was trying to stay at the requirements level to start with, but:
- I'd already suggested to @EduardMe that we need an enumerator for windows (like you've shown)
- in an earlier draft of this I included a request to close Windows too, but removed it as a lower priority ask

To continue at the same level, then, there's probably a simpler first step. That's to extend `Editor.openNoteByFilename()` further, to add a parameter `reuseExistingWindow?: boolean` which defaults to true.

**Comment by dwertheimer** [comment]
type === main|split|floating|html

**Comment by dwertheimer** [comment]
when you open a window using the API there should be an optional param that you give it a name so you can know which it is later
